Bootloader: Parse ELF headers to find pydata section

durko opened the pull request:
This PR replaces the "search for cookie from end of file" approach with a minimal ELF parser to find the pydata section.


## Rationale

On Linux the python payload is embedded in a custom ELF section called `pydata`. On execution the bootloader has to find this section in the binary.

Currently booloader searches the last 4096 + sizeof(COOKIE) bytes of the file for a magic string. This works sufficiently well under most circumstances, as the only section following the `pydata` section will be `.shstrtab`, which is usually only a few hundred bytes in size.

When using [staticx](https://github.com/JonathonReinhart/staticx) on a pyinstaller binary, staticx runs [patchelf](https://github.com/NixOS/patchelf) to adjust interpreter and rpath. Patchelf (sometimes) reorders the ELF sections of a binary. Depending on the amount of data ending up behind the pydata section the COOKIE can easily fall out of search range.

We spotted this problem on first on Ubuntu aarch64 JonathonReinhart/staticx#93.

Fixes #3441

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -235,14 +235,15 @@ pyi_arch_extract2fs(ARCHIVE_STATUS *status, TOC *ptoc)
 #if defined(WIN32)
 #define SEARCH_SIZE (8 + sizeof(COOKIE))
-#else
+#elif defined(__APPLE__)
```
Why not special case Linux instead? This will also affect other platforms, like Solaris or AIX.

durko replied:
AIX, HPUX, Solaris are all also ELF systems. Does PyInstaller embed the payload differently on other Unixes? The comment does not single out the other platforms.

I haven't looked into the embedding process up to now. On quick glance `binding/api.py` gives me the impression that data is appended to the end of the binary on non-linux Unixes. Only OSX gets the additional fixup call for the MachO binary. That would place the payload and cookie right at the end of the binary, which would render the above SEARCH_SIZE correct.

lkollar replied:
I suppose other platforms which use ELF should work too. AIX uses XCOFF though, so it won't work there.

htgoebel replied:
I wonder whether there is some precompiler symbol telling if ELF is used as output format?

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -398,6 +538,8 @@ pyi_arch_open(ARCHIVE_STATUS *status)
 #if defined(WIN32) || defined(__APPLE__)
     search_end = findDigitalSignature(status);
+#else
+    search_end = findElfSection(status);
```
Similar as at the top of the file: we assume that all platforms except Windows and macOS use ELF and work the same way as Linux. I think it's safer to have a check just for Linux and call the function only there.

durko replied:
See above. Is it incorrect to assume ELF is being used on the other Unixes?

Note: If the pydata section is not found by the function call, it falls back to the search from end of file here.

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+static int
+findElfSection(ARCHIVE_STATUS * const status)
+{
+#if !defined(WIN32) && !defined(__APPLE__)
```
This seems unnecessary as there is already a platform check macro around the function call.

durko replied:
The conditionals here mirror the respective conditionals in `findDigitalSignature`.

@htgoebel I'll remove those and check if it compiles on windows and apple.

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+{
+#if !defined(WIN32) && !defined(__APPLE__)
+    char head[64];
+    uint16_t endianess_bytes = 0x0201;
```
Isn't endianness a single byte?

durko replied:
@lkollar @htgoebel Reading one byte of a multibyte int is a common runtime check for endianess.

The EI_DATA field in the ELF header is a byte that indicates little (value=1) or big (value=2) endian. Putting `endianess_bytes` with the 16bit value 0x0201 on the stack and reading its first byte yields 0x01 on little and 0x02 on big endian systems.

We can #define the value elsewhere but we have to put it into an addressable memory location here to access a single byte.

lkollar replied:
@durko sorry, totally missed this one. Feel free to dismiss my comments around the endianness check.

htgoebel replied:
Regarding the whole endiness checks: I'm in favor of less "magic", thus someone less skill in C (like me ;-) is able to understand the code easily.

I suggest using more verbose code, something along these lines:
```
// these values need to match ELF header flags
#define BIG_ENDIAN 2    # or ENDIAN_BIG, I don't care
#define LITTLE_ENDIAN 1
// commonly used way to determine CPU endianess: Put uint16 onto the
// stack and read the first byte to learn whether it is the upper or lower byte.
uint16_t endianess_bytes = BIG_ENDIAN <<8 + LITTLE_ENDIAN;
uint8_t endianess = *(uint8_t*)&endianess_bytes;
…
if (head[0x05] != endianess {
…
```

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    }
+
+    /* Ensure endianess of file matches */
+    if (head[0x05] != *(uint8_t*)&endianess_bytes) {
```
This will convert `endianess_bytes` to 1. You don't need to cast if you only use a single byte when you declare the variable.

durko replied:
@lkollar Depending on endianess we read 0x01 or 0x02 here and compare with the header value. See also above.

@htgoebel Not sure if it could be an issue on e.g. bi-endian systems. Since it is an inexpensive check, I decided to include it for safety. What do you think?

htgoebel replied:
ACK for the test. But is there some symbol instead of using `0x05`? If there is none in a header-file, I suggest defining one ourself. (If there is a header file, you could still consider copying the definition here.)

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+
+    /* Ensure endianess of file matches */
+    if (head[0x05] != *(uint8_t*)&endianess_bytes) {
+        VS("LOADER: %s endianess does not match CPU\n", status->archivename);
```
This assumes that the CPU is little-endian. I don't know if PyInstaller works correctly on big-endian platforms at the  moment, but this should either check the endianness of the CPU or simply say that big-endian ELF files are not supported.

durko replied:
@lkollar See above. It actually does work on both little and big endian ARM systems.

lkollar commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    }
+
+    /* Ensure ELF version is supported */
+    if (head[0x06] != 1 || *(uint32_t*)(head + 0x14) != 1) {
```
Why not use array notation in the second comparison to make it consistent:

`if (head[0x06] != 1 || head[0x14]) != 1) {`

htgoebel replied:
@durko I suggest using a tighter check along these lines:
```
if ((endianess == LITTLE_ENDIAN && head[0x06] != 1) || 
   ((endianess == BIG_ENDIAN && *(uint32_t*)(head + 0x14) != 1)) {
```
(And again, please use some symbols for 00x5 and 0x149

htgoebel commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+        free(section);
+        free(shstr);
+        VS("LOADER: %s could not find pydata section\n", status->archivename);
+        return -1;
```
This code in repeated quite often. What dou you think about using some (ugly) `goto fail` and having this code only once? Please have a look how this is done in the other parts of the bootloader and implement it the same way.

htgoebel commented on bootloader/src/pyi_archive.c:
```diff
@@ -235,14 +235,15 @@ pyi_arch_extract2fs(ARCHIVE_STATUS *status, TOC *ptoc)
- * move the ELF section headers so they follow the cookie, so we need to search backward
- * past the section headers to find the cookie.
+ * we use objtool to insert the archive into the bootloader and on runtime parse the
+ * ELF headers to extract the precise end offset.
```
The backward search is still in for non ELF files, isn't it? Thus the comment should state this.

durko replied:
@htgoebel Going through the feedback here today I have come to the understanding that other Unixes might append the payload at the end even if their file format is also ELF. If there is no fixup or signing as on MachO binaries on OSX, is a backward search necessary on other Unixes or is the data right there at the end?

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -235,14 +235,15 @@ pyi_arch_extract2fs(ARCHIVE_STATUS *status, TOC *ptoc)
+#elif defined(__APPLE__)
 #define SEARCH_SIZE (4096 + sizeof(COOKIE))
+#else
+#define SEARCH_SIZE (sizeof(COOKIE))
```
@htgoebel There is no uncertainty anymore when parsing the ELF header, we do know the exact location of the cookie.

htgoebel replied:
ACK, but if reading the .pydata section fails, the code falls back to dump search, where the smaller size might cause problems as soon as the cookie is not at the very end of the file.

Maybe pass `search_size` as an parameter to `pyi_arch_find_cookie()`, which might vary if we have an ELF section?

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    char *shstr;
+    uint32_t section_idx;
+    uint64_t pydata_offset;
+    uint64_t pydata_size;
```
ELF headers differ for 32bit and 64 bit binaries and define fixed bit-length fields for the different header values regardless of the target platform. So `long` can be different size on two platforms using a e.g. 32bit ELF header, but the header fields will be same width. 

 All values we read or derive from header content should be fixed with. For example `pydata_size` takes a value directly from a 64bit header filed, while `section_idx` is a counter that needs to be able to hold values up until a number that is defined by a 32bit header field.

htgoebel replied:
IC, all these variables are read directly from the file.

`chtr` was a typo, should have been `shstr`.

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    }
+
+    /* Check if ELF magic present */
+    if (head[0] != 0x7F || head[1] != 'E' || head[2] != 'L' || head[3] != 'F') {
```
@htgoebel For endianess reasons we have to compare byte by byte

htgoebel replied:
ACK, Please write this into the comment.

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+        e_shoff = *(uint32_t*)(head + 0x20);
+        e_shentsize = *(uint16_t*)(head + 0x2e);
+        e_shnum = *(uint16_t*)(head + 0x30);
+        e_shstrndx = *(uint16_t*)(head + 0x32);
```
@htgoebel We could look into using headers. My first choice using the well known offsets was to avoid having to ifdef a bunch of headers for Linux and other Unix platforms. Without the code could compile on Windows/Apple. Regarding the ifdef comments above, what would be your preference?

At least on Linux we have Multilib, which allows us to execute 32bit binaries on a 64bit kernel and userland.

htgoebel replied:
IC, all these values are read from the file.

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    if (section == NULL) {
+        FATAL_PERROR("malloc", "Could not allocate buffer for section header.");
+        return -1;
+    }
```
@htgoebel In principle this value comes from a 16bit field. Realistically it is smaller under normal circumstances. We could have fixed sized memory block on the stack, but would still have to check if that is enough on runtime and revert to heap allocations or fail the parsing.

htgoebel replied:
This is too much hazzle. Let's stay with `malloc` ad you did. If we are out of memory her alreade, the program will never have enough for running anyway :-)

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+        free(section);
+        FATAL_PERROR("malloc", "Could not allocate buffer for ELF shstrs.");
+        return -1;
+    }
```
See above

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+    } else {
+        shstr_offset = *(uint32_t*)(section + 0x10);
+        shstr_size = *(uint32_t*)(section + 0x14);
+    }
```
See above

durko commented on bootloader/src/pyi_archive.c:
```diff
@@ -377,6 +378,145 @@ findDigitalSignature(ARCHIVE_STATUS * const status)
+
+    /* Load ELF section header parameters */
+    bits = head[0x04]; /* 2=64, 1=32 */
+    if (bits == 2) {
```
@htgoebel Yes will do, that will make it more readable

durko commented:
@lkollar @htgoebel Added a fixup commit https://github.com/pyinstaller/pyinstaller/commit/d16300028a88bedbe61ebe067ab8f57c314f94e9 addressing the points discussed above. The branch is rebased onto develop with the fixup commit autosquashed into the original commit.

I've went through all issues mentioning the bootloader, ELF or the error messages generated when the payload is not found. Issue #3441 should get closed with this change.

htgoebel commented:
@durko Thanks for updating so quickly and for you explanations! Most of the code is okay now, I'd just like to discuss/solve these points:

* How to determine whether the file format is expected to be ELF?

  I'd prefer to use this new code on all platforms supporting ELF. OTOH PyInstaller itself supports the `.pydata` section for Linux only: https://github.com/pyinstaller/pyinstaller/blob/v3.5/PyInstaller/building/api.py#L598

  I suggest:

  1. Make the new code decide on some `ELF_FORMAT_AVAILABLE` symbol. For now this is True of we are on linux, and False on any other platform.

  2. Add a comment to this define stateing that `.pydata` section is currently only supported on Linux, referencing to building/api.py.

  3. Open two new issues:
      - Make PyInstaller use `objcopy` an more platforms. (Done, see  #4528)
      - Make wscript determine if the executables will be build in ELF, write this into some config.h and make bootloader/src/pyi_archive.c use it.

* Header files or not: Im in favor of at least using symbols instead of numeric offsets. I'd prefer to use header files, and at top of this file there are already conditional includes. Though this might cause failures if the resp. header files are not available on some platform for any reasons. What do you think?

* Please rethink SEARCH_SIZE (see inline comments)

htgoebel commented:
@durko Any update on this? I plan to to a last release supporting 2.7 before or shortly after 2020-01-01, which i would like to have this change in.
